drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
+        let header = Header::new(msg_id, response.kind());
+
+        let hdr_len = hubpack::serialize(tx_data_buf, &header).unwrap();
+        let msg_len =
+            hubpack::serialize(&mut tx_data_buf[hdr_len..], &response).unwrap();
+
+        ringbuf_entry!(Trace::ResponseSize(ResponseSize {
+            header_length: hdr_len as u8,
+            message_length: msg_len as u8,
+            data_length: data_len as u16
+        }));
 
         // At this point, any supplementary data was written to
         // `tx_data_buf[Message::MAX_SIZE..]`, so it's not necessarily tightly
         // packed against the end of the `Message`.  Let's shift it backwards
         // based on the size of the leading `Message`:
         tx_data_buf.copy_within(
-            Message::MAX_SIZE..(Message::MAX_SIZE + data_len),
-            msg_len,
+            reserved_framing..(reserved_framing + data_len),
+            hdr_len + msg_len,
         );
-        Some((msg_len + data_len) as u32)
+        Some((hdr_len + msg_len + data_len) as u32)
     }
 
     fn handle_host_request(
         &mut self,
         h: HostRequest,
-        modules: ModuleId,
         data: &[u8],
         out: &mut [u8],
-    ) -> Result<(SpResponse, usize), Error> {
+    ) -> (MessageBody, usize) {
         match h {
-            HostRequest::Status => {
+            HostRequest::Status(modules) => {
                 ringbuf_entry!(Trace::Status(modules));
-                let count = self.get_status(modules, out)?;
-                Ok((SpResponse::Status, count))
+                let mask = LogicalPortMask(modules.0 as u32);
+                let (num_status_bytes, result) = self.get_status(mask, out);
+                ringbuf_entry!(Trace::OperationResult(result));
+                let success = ModuleId(result.success().0 as u64);
+                let (final_payload_len, failed_modules) = self.handle_errors(
+                    modules,
+                    result,
+                    None,
+                    num_status_bytes,
+                    out,
+                );
+
+                (
+                    MessageBody::SpResponse(SpResponse::Status {
+                        modules: success,
+                        failed_modules,
+                    }),
+                    final_payload_len,
+                )
             }
-            HostRequest::Read(mem) => {
-                ringbuf_entry!(Trace::Read(modules, mem));
-                let out_size = mem.len() as u32 * modules.ports.0.count_ones();
-                if out_size as usize > transceiver_messages::MAX_PAYLOAD_SIZE {
-                    return Err(Error::RequestTooLarge);
+            HostRequest::Read { modules, read } => {
+                ringbuf_entry!(Trace::Read(modules, read));
+                let mask = LogicalPortMask(modules.0 as u32);
+                let request_size = read.len() as u32 * mask.0.count_ones();
+                if request_size as usize
```

Comment:
`request_size` can be retrieved as a `usize` already using `HostRequest::expected_data_len()`, if that's helpful!

Reply:
That isn't helpful while handling a `HostRequest::Read` since it is just meant to return the amount of data to write via `HostRequest::Write`.